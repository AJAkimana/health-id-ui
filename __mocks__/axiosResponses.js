export const resolvedRequest = {
  status: 200,
  response: {
    data: {},
    statusText: 'OK',
    headers: {},
    config: {}
  }
};

export const rejectedRequest = {
  status: 400,
  response: {
    data: { errors: { message: 'errors' } },
    statusText: 'ERROR',
    headers: {},
    config: {}
  }
};
