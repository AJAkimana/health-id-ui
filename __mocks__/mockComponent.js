import React from 'react';

const MockComponent = () => <div />;

export default MockComponent;
