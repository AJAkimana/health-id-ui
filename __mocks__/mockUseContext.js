import * as StateContext from '../src/providers/stateProvider';

const contextValues = [{ isActive: 'grid1' }, jest.fn()];
const contextMock = jest
  .spyOn(StateContext, 'useStateValue')
  .mockImplementation(() => contextValues);
export default contextMock;
