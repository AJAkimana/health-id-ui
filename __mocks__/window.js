const map = {};
window.URL.createObjectURL = jest.fn(file => file);

window.document.addEventListener = jest.fn((event, callback) => {
  map[event] = callback;
});
window.addEventListener = jest.fn((event, callback) => {
  map[event] = callback;
});

const event = jest.fn(e => map[e.name](e));
export const document = { event };
export class MockFileReader extends FileReader {
  constructor() {
    super();
    return {
      removeEventListener: jest.fn((e, callback) => {
        map[event] = callback;
      }),
      addEventListener: jest.fn((e, callback) => {
        map[event] = callback;
      }),
      readAsDataURL: jest.fn(),
      result: {}
    };
  }
}
window.FileReader = MockFileReader;
export default {
  ...document,
  document,
  FileReader,
  map
};
