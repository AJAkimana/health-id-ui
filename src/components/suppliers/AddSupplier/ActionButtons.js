import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { Button } from '@material-ui/core';
import { ActionButtonStyles } from '../../../assets/styles/suppliers/addSupplierStyles';


const ActionButtons = (props) => {
  const { handleSendForApproval, handleAddAnotherSupplier, disabled } = props;

  return (
    <Fragment>
      <Button
        id="saveNew"
        disabled={disabled}
        variant="outlined"
        style={ActionButtonStyles.doneButton}
        className="new-btn"
        onClick={handleAddAnotherSupplier}
      >
      SAVE AND ADD NEW
      </Button>
      <Button
        id="save"
        disabled={disabled}
        variant="contained"
        color="primary"
        style={ActionButtonStyles.addButton}
        className="create-btn"
        onClick={handleSendForApproval}
      >
    SAVE
      </Button>
    </Fragment>
  );
};

ActionButtons.propTypes = {
  handleAddAnotherSupplier: PropTypes.func.isRequired,
  handleSendForApproval: PropTypes.func.isRequired,
  disabled: PropTypes.bool,
};

ActionButtons.defaultProps = {
  disabled: false
};

export default ActionButtons;
