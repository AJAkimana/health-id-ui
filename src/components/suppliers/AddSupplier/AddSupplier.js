/* eslint-disable class-methods-use-this */
/* eslint-disable react/sort-comp */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import { compose, graphql, Query } from 'react-apollo';
import { withRouter } from 'react-router-dom';
import {
  validateEmail, validatePhone, validateName, validateComment
} from '../../utils/validations';
import SupplierForm from './SupplierForm';
import withAuth from '../../withAuth';
import GET_INITIAL_DATA from '../../../queries/countryQuery';
import CREATE_SUPPLIER from '../../../mutations/createSupplier';
import BackAction from '../../shared/BackAction';
import notify from '../../shared/Toaster';
import verifyFile from '../../../utils/products/verifyFile';
import FormLoader from '../../shared/Loader/FormLoader';

import { StateContext } from '../../../providers/stateProvider';

export class AddSupplier extends Component {
  state = {
    name: '',
    nameError: '',
    nameHelperText: '',
    email: '',
    emailError: false,
    emailHelperText: '',
    comment: '',
    commentError: '',
    commentHelperText: '',
    lineError: '',
    lineHelperText: '',
    mobileNumber: '',
    mobileNumberError: false,
    mobileHelperText: '',
    addressLine1: '',
    addressLine2: '',
    lga: '',
    commentary: '',
    countryValue: '',
    cities: '',
    cityValue: '',
    cityId: '',
    tierId: '',
    countryId: '',
    logo: '',
    paymentTermsId: '1',
    creditDays: 0,
    loading: false,
    imageFile: '',
    fileName: '',
    src: null,
    crop: {
      aspect: 1 / 1
    },
    open: false,
    isDisabled: true,
    colorHasChanged: false,
    colorHasChangedCity: false
  };

  componentDidMount() {
    const [, dispatch] = Object.values(this.context);
    dispatch({
      type: 'changeGrid',
      grid: 'grid4'
    });
  }

  handleSliderChange = (event) => {
    if (event === 0) {
      this.setState({ creditDays: event });
      this.setState({ paymentTermsId: '1' });
    } else {
      this.setState({ creditDays: event });
      this.setState({ paymentTermsId: '2' });
    }
  }

  handleNameChange = (event) => {
    const { name, value } = event.target;
    const isValid = validateName(value);
    const [helperText, error] = isValid;
    this.setState({
      [name]: value,
      nameError: error,
      nameHelperText: helperText
    });
  };

  handleCommentChange = (event) => {
    const { value } = event.target;

    const isValid = validateComment(value);
    const [helperText, error] = isValid;
    this.setState({
      commentary: value,
      commentError: error,
      commentHelperText: helperText
    });
  };

  handleLineChange = (event) => {
    const { name, value } = event.target;

    const isValid = validateName(value);
    const [helperText, error] = isValid;
    this.setState({
      [name]: value,
      lineError: error,
      lineHelperText: helperText
    });
  }

  handleChange = (event) => {
    const { name, value } = event.target;
    this.setState({
      [name]: value,
    });
  }

  handleCountryChange = (event) => {
    const {
      label, value, citySet, id
    } = event;
    this.setState({
      colorHasChanged: true,
      countryValue: { label, value },
      cities: citySet,
      cityValue: {
        label: citySet[0].name,
        value: citySet[0].name
      },
      cityId: citySet[0].id,
      countryId: id
    });
  };

  handleCityChange = (event) => {
    const { label, value, id } = event;
    this.setState({
      cityValue: { label, value },
      cityId: id
    });
  };

  handlePaymentTermsChange = (event) => {
    const { value } = event.target;
    this.setState({ paymentTermsId: value });
  };

  handleMobileChange = (value) => {
    const isValid = validatePhone(value);
    const [helperText, error] = isValid;
    this.setState({
      mobileNumber: value,
      mobileNumberError: error,
      mobileHelperText: helperText
    });
  };

  handleEmailChange = (event) => {
    const { value } = event.target;
    const isValid = validateEmail(value);
    const [helperText, error] = isValid;
    this.setState({
      email: value,
      emailError: error,
      emailHelperText: helperText
    });
  };

  handleColorChange = () => {
    this.setState({ colorHasChanged: true });
  }

  handleColorChangeCity = () => {
    this.setState({ colorHasChangedCity: true });
  }


  handleTierChange = (event) => {
    const { value } = event.target;
    this.setState({ tierId: value });
  };

  handleProposeSupplier = (btnClicked) => {
    const { addSupplier } = this.props;
    const {
      name,
      cityId,
      email,
      mobileNumber,
      addressLine1,
      addressLine2,
      lga,
      paymentTermsId,
      commentary,
      tierId,
      countryId,
      creditDays,
      logo
    } = this.state;
    const paymentsTermsId = parseInt(paymentTermsId, 10);
    const citId = parseInt(cityId, 10);
    this.setState({ loading: true });
    addSupplier({
      variables: {
        name,
        email,
        mobileNumber,
        addressLine1,
        addressLine2,
        lga,
        paymentsTermsId,
        commentary,
        citId,
        tierId,
        countryId: parseInt(countryId, 10),
        creditDays,
        logo: logo || 'none'
      }
    })
      .then((res) => {
        const { history, refetch } = this.props;
        const { addSupplier: addedSupplier = {} } = res.data;
        const { name: supplierName = {}, supplierId } = addedSupplier.supplier;
        notify(`${supplierName} with id ${supplierId} has been added and sent for approval`);

        if (btnClicked === 'save') {
          history.push('/suppliers');
          refetch();
        }
      })
      .catch((err) => {
        const { message } = err.graphQLErrors[0];
        notify(message);
      });
  }

  handleSendForApproval = async (e) => {
    const btnClicked = e.currentTarget.id;
    await this.handleProposeSupplier(btnClicked);
  }

  handleAddAnotherSupplier = async (e) => {
    const btnClicked = e.currentTarget.id;
    await this.handleProposeSupplier(btnClicked);
    setTimeout(() => {
      this.setState({
        name: '',
        email: '',
        mobileNumber: '',
        addressLine1: '',
        addressLine2: '',
        lga: '',
        commentary: '',
        cityId: '',
        countryValue: '',
        cities: '',
        tierId: '',
        countryId: '',
        logo: '',
        paymentTermsId: '1',
        creditDays: '',
        loading: false,
        imageFile: '',
        fileName: '',
        src: null,
        crop: {
          aspect: 1 / 1
        },
        open: false,
      });
    }, 1500);
  }

  onSelectFile = (e) => {
    const { files } = e.target;
    const imageFile = e.target.files[0];
    const maxFileSize = 1000000; // bytes
    const acceptedFileType = 'image/jpg, image/jpeg, image/JPEG, image/png, image/PNG';
    if (files && files.length > 0) {
      const verified = verifyFile(files, maxFileSize, acceptedFileType);
      if (verified) {
        this.setState({
          fileName: files[0].name,
          imageFile
        });

        const reader = new FileReader();
        reader.addEventListener('load', () => {
          this.setState({
            src: reader.result,
            open: true
          });
        }, false);
        reader.readAsDataURL(imageFile);
      }
    }
  }

  getCroppedImg = (imageFile, pixelCrop, fileName) => {
    const canvas = document.createElement('canvas');
    canvas.width = pixelCrop.width;
    canvas.height = pixelCrop.height;
    const ctx = canvas.getContext('2d');

    const image = new Image();
    const promise = new Promise((resolve) => {
      image.onload = (() => {
        ctx.drawImage(
          image,
          pixelCrop.x,
          pixelCrop.y,
          pixelCrop.width,
          pixelCrop.height,
          0,
          0,
          pixelCrop.width,
          pixelCrop.height
        );
        resolve();
      });
      image.src = imageFile;
    }).then(() => new Promise((resolve) => {
      canvas.toBlob((blob) => {
        blob.name = fileName;
        resolve(blob);
      }, 'image/jpeg');
    }));
    return promise;
  }

  handleImageDrop = (file) => {
    const formData = new FormData();
    formData.append('file', file);
    formData.append('upload_preset', process.env.UPLOAD_PRESET);
    formData.append('api_key', `${process.env.API_KEY}`);
    formData.append('timestamp', (Date.now() / 1000) || 0);
    return axios({
      url: process.env.CLOUDINARY_URL,
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded'
      },
      data: formData
    }).then((response) => {
      const { data } = response;
      const fileURL = data.secure_url;
      this.setState({
        logo: fileURL
      });
      notify('Image uploaded successfully');
    }).catch((err) => {
      notify('There was an error uploading the image', err);
    });
  }

  handleDragOverImage = (acceptedFiles) => {
    const reader = new FileReader();
    reader.addEventListener('load', () => {
      const binaryStr = reader.result;
      this.setState({ imageFile: acceptedFiles, src: binaryStr, open: true });
    }, false);

    acceptedFiles.forEach(file => reader.readAsDataURL(file));
  }

  handleSave = () => {
    const {
      src,
      fileName,
      crop
    } = this.state;

    this.getCroppedImg(src, crop, fileName).then((data) => {
      this.handleImageDrop(data);
      this.setState({
        src: '',
        open: false
      });
    });
  }

  handleClose = () => {
    const { src } = this.state;
    this.setState({
      src: '',
      open: false
    });
    this.handleImageDrop(src);
  }


  handleOnCropChange = (crop) => {
    this.setState({ crop });
  }

  static contextType = StateContext;

  render() {
    return (
      <div>
        <Query
          query={GET_INITIAL_DATA}
          variables={{ outletId: localStorage.outletId }}
        >
          {({ loading, error, data }) => (loading && <FormLoader />) || (error && <p>Error</p>) || (
            <div>
              <BackAction
                header="Add Supplier"
                link="/suppliers"
              />
              <SupplierForm
                state={this.state}
                initialData={data}
                handleChange={this.handleChange}
                handleNameChange={this.handleNameChange}
                handleCommentChange={this.handleCommentChange}
                handleLineChange={this.handleLineChange}
                onSelectFile={this.onSelectFile}
                handleOnDrop={this.handleImageDrop}
                handleOnCropChange={this.handleOnCropChange}
                handleTierChange={this.handleTierChange}
                handleCountryChange={this.handleCountryChange}
                handleCityChange={this.handleCityChange}
                handleClose={this.handleClose}
                handleSave={this.handleSave}
                handleSendForApproval={this.handleSendForApproval}
                handleAddAnotherSupplier={this.handleAddAnotherSupplier}
                handlePaymentTermsChange={this.handlePaymentTermsChange}
                handleMobileChange={this.handleMobileChange}
                handleEmailChange={this.handleEmailChange}
                handleColorChange={this.handleColorChange}
                handleColorChangeCity={this.handleColorChangeCity}
                handleSliderChange={this.handleSliderChange}
                handleDragImage={this.handleDragOverImage}
              />
            </div>
          )}
        </Query>
      </div>
    );
  }
}

AddSupplier.propTypes = {
  history: PropTypes.objectOf(PropTypes.any),
  addSupplier: PropTypes.func.isRequired,
  refetch: PropTypes.func,
};

AddSupplier.defaultProps = {
  history: {},
  refetch: () => { }
};

const ADD_SUPPLIER = graphql(CREATE_SUPPLIER, { name: 'addSupplier' });

export default withAuth((compose(
  ADD_SUPPLIER
)(withRouter(AddSupplier))));
