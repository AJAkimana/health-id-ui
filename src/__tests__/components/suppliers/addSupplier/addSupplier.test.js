import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import * as moxios from 'moxios';
import { AddSupplier } from '../../../../components/suppliers/AddSupplier/AddSupplier';
import GET_INITIAL_DATA from '../../../../queries/citiesCreditdaysTierQuery';
import '../../../../../__mocks__/window';
import { resolvedRequest, rejectedRequest } from '../../../../../__mocks__/axiosResponses';

describe('Render Add Supplier component', () => {
  jest.useFakeTimers();
  beforeEach(() => {
    moxios.install();
  });
  afterEach(() => {
    moxios.uninstall();
  });
  //   class MockFileReader {
  //     onerror() { }
  //     onload() { }
  //     readAsDataURL() {
  //         this.result = 'result';
  //         this.onload();
  //     }
  // }
  // beforeEach(() => {
  //   (window as WindowWithFileReader).FileReader = MockFileReader;
  //   timer.mock();
  // });

  // afterEach(() => {
  //   (window as WindowWithFileReader).FileReader = originalFileReader;
  //   timer.restore();
  // });

  const notify = jest.fn();
  const props = {
    session: {
      me: {
        mobileNumber: '254123123123',
        email: 'user@gmail.com',
        username: 'user',
        role: { name: 'Master Admin' },
        activeOutlet: {
          outletpreference: {
            outletTimezone: { name: 'Africa/Lagos' }
          }
        }
      }
    },
    addSupplier: jest.fn(
      () => new Promise((resolve) => {
        const data = {
          addSupplier: {
            supplier: {
              name: 'name',
              supplierId: 1
            }
          }
        };
        const res = { data };
        return resolve(res);
      })
    ),
    getAttribute: jest.fn()
  };

  const mocks = [
    {
      request: {
        query: GET_INITIAL_DATA
      },
      result: {
        data: {
          cities: [
            {
              id: '16',
              name: 'Abuja'
            },
            {
              id: '17',
              name: 'Lagos'
            },
            {
              id: '18',
              name: 'Ido'
            },
            {
              id: '19',
              name: 'Ido'
            },
            {
              id: '20',
              name: 'Lagos'
            }
          ]
        }
      }
    }
  ];

  const crop = {
    x: 0,
    y: 10,
    width: 200,
    height: 200,
    aspect: 1 / 1
  };

  const wrapper = shallow(<AddSupplier {...props} />);

  const validFile = [
    {
      name: 'eucerin',
      size: 10000,
      type: 'image/jpg',
      lastModified: ''
    }
  ];

  const invalidFile = [
    {
      name: 'eucerin',
      size: 10000,
      type: 'text/csv',
      lastModified: ''
    }
  ];

  const data = [
    {
      name: 'eucerin',
      size: 10000,
      type: 'image/jpg',
      lastModified: ''
    }
  ];

  const largeFile = [
    {
      name: 'eucerin',
      size: 1000000000,
      type: 'image/jpg',
      lastModified: ''
    }
  ];

  const fileName = 'eucerin';

  it('renders data upon successful view', () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <BrowserRouter>
          <AddSupplier {...props} />
        </BrowserRouter>
      </MockedProvider>
    );
    wait(0);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls handle slider change function', () => {
    const attr = { value: undefined };
    const event = {
      target: {
        getAttribute: value => attr[value]
      }
    };
    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.instance().handleSliderChange(event);
    expect(wrapper.state().creditDays).toEqual(event);
  });
  it('calls handle change name function', () => {
    const event = {
      target: {
        name: 'name',
        value: 'Ema'
      }
    };

    const value = 'Ema';

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.instance().handleNameChange(event);
    expect(wrapper.state().name).toEqual(value);
  });
  it('calls handle change comment function', () => {
    const event = {
      target: {
        name: '',
        value: 'comment'
      }
    };

    const value = '';

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.instance().handleCommentChange(event);
    expect(wrapper.state().name).toEqual(value);
  });
  it('calls handle change line function', () => {
    const event = {
      target: {
        name: 'name',
        value: 'House 12 Binta road'
      }
    };

    const value = 'House 12 Binta road';

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.instance().handleLineChange(event);
    expect(wrapper.state().name).toEqual(value);
  });
  it('calls handle color change function', () => {
    const event = {};
    const colorHasChanged = true;

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.instance().handleColorChange(event);
    expect(wrapper.state().colorHasChanged).toEqual(colorHasChanged);
  });
  it('calls handle color city change function', () => {
    const event = {};
    const colorHasChangedCity = true;

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.instance().handleColorChangeCity(event);
    expect(wrapper.state().colorHasChangedCity).toEqual(colorHasChangedCity);
  });

  it('calls handle tier change function', () => {
    const event = {
      target: {
        value: 1
      }
    };

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.setState({ tierId: event.target.value });
    wrapper.instance().handleTierChange(event);

    expect(wrapper.instance().state.tierId).toBe(1);
  });

  it('calls handle payment terms change function', () => {
    const event = {
      target: {
        value: 1
      }
    };

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.setState({ paymentTermsId: event.target.value });
    wrapper.instance().handlePaymentTermsChange(event);

    expect(wrapper.instance().state.paymentTermsId).toBe(1);
  });

  it('calls handle country change function', () => {
    const event = {
      value: 'Nigeria',
      label: 'Nigeria',
      citySet: [
        {
          id: 22,
          name: 'Nairobi'
        },
        {
          id: 23,
          name: 'Mombasa'
        },
        {
          id: 24,
          name: 'Kitui'
        }
      ]
    };

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.setState({
      countryValue: {
        label: event.label,
        value: event.value
      },
      cities: event.citySet
    });
    wrapper.instance().handleCountryChange(event);

    expect(wrapper.instance().state.cityId).toBe(22);
    expect(wrapper.instance().state.cityValue.label).toBe('Nairobi');
  });

  it('calls handle city change function', () => {
    const event = {
      label: 'Lagos',
      value: 'Lagos',
      id: 2
    };

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.setState({
      cityValue: {
        label: event.label,
        value: event.value
      },
      cityId: event.id
    });
    wrapper.instance().handleCityChange(event);

    expect(wrapper.instance().state.cityId).toBe(2);
    expect(wrapper.instance().state.cityValue.label).toBe('Lagos');
  });

  it('calls handle email change function and return error when email is not correct', () => {
    const event = {
      target: {
        value: 'email'
      }
    };

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.setState({ email: event.target.value });
    wrapper.instance().handleEmailChange(event);

    expect(wrapper.instance().state.email).toBe('email');
    expect(wrapper.instance().state.emailError).toBe(true);
    expect(wrapper.instance().state.emailHelperText).toBe('Invalid Email');
  });

  it('calls handle email change function and return no error when email is correct', () => {
    const event = {
      target: {
        value: 'email@email.com'
      }
    };

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.setState({ email: event.target.value });
    wrapper.instance().handleEmailChange(event);

    expect(wrapper.instance().state.email).toBe('email@email.com');
    expect(wrapper.instance().state.emailError).toBe(false);
    expect(wrapper.instance().state.emailHelperText).toBe('');
  });

  it('calls handle mobile change function and return no error when number is correct', () => {
    const value = '+2347030303030';

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.setState({ mobileNumber: value });
    wrapper.instance().handleMobileChange(value);

    expect(wrapper.instance().state.mobileNumber).toBe('+2347030303030');
    expect(wrapper.instance().state.mobileNumberError).toBe(false);
    expect(wrapper.instance().state.mobileHelperText).toBe('');
  });

  it('calls handle mobile change function and return error when number is inccorrect', () => {
    const value = '7030303030';

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.setState({ mobileNumber: value });
    wrapper.instance().handleMobileChange(value);

    expect(wrapper.instance().state.mobileNumber).toBe('7030303030');
    expect(wrapper.instance().state.mobileNumberError).toBe(true);
    expect(wrapper.instance().state.mobileHelperText).toBe('Invalid Phone Number');
  });

  it('calls handlePropose supplier', () => {
    const res = {
      data: {
        addedSupplier: {
          supplierName: 'suupplier',
          supplierId: 1
        }
      }
    };
    const btnClicked = 'save';
    const message = 'test';
    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.instance().handleProposeSupplier(btnClicked);
    notify(message);
    expect(wrapper.state().loading).toBeTruthy();
  });

  it('calls handleSendForApproval', () => {
    const event = {
      currentTarget: {
        id: 'save'
      }
    };

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.instance().handleSendForApproval(event);
    // expect(wrapper.state().loading).toBeTruthy();
  });

  it('calls handleAddAnotherSupplier', () => {
    const event = {
      currentTarget: {
        id: 'saveNew'
      }
    };
    const btnClicked = 'save';

    const wrapper = shallow(<AddSupplier {...props} />);
    wrapper.instance().handleProposeSupplier(btnClicked);
    wrapper.instance().handleAddAnotherSupplier(event);
    expect(setTimeout).toHaveBeenCalledTimes(6);
    // expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1500);
    // expect(wrapper.state().loading).toBeTruthy();
  });

  it('calls onSelectFile function on a large image file', () => {
    const e = {
      target: {
        files: largeFile
      }
    };
    wrapper.instance().onSelectFile(e);
  });

  it('calls handle drag image function ', () => {
    const reader = {
      result: 'goodFile'
    };

    const spy = jest.spyOn(wrapper.instance(), 'handleDragOverImage');
    const acceptedFiles = [new Blob()];
    wrapper.instance().handleDragOverImage(acceptedFiles);
    expect(spy).toHaveBeenCalled();
  });
  it('calls onSelectFile function on a non-image file', () => {
    const e = {
      target: {
        files: invalidFile
      }
    };
    const spy = jest.spyOn(wrapper.instance(), 'onSelectFile');
    wrapper.instance().onSelectFile(e);
    console.log('wrapper.state().src', wrapper.state().src);
    expect(spy).toHaveBeenCalled();
  });
  it('calls onSelectFile function on a image file', () => {
    const validFile = new File([new Blob()], 'image.jpg', {
      name: 'profile',
      size: 10000,
      type: 'image/jpg',
      lastModified: ''
    });
    const e = {
      target: {
        files: [validFile]
      }
    };
    const spy = jest.spyOn(wrapper.instance(), 'onSelectFile');
    wrapper.instance().onSelectFile(e);
    console.log('wrapper.state().src', wrapper.state().src);
    expect(spy).toHaveBeenCalled();
  });

  it('calls getCroppedImage function', () => {
    const spy = jest.spyOn(wrapper.instance(), 'getCroppedImg');
    wrapper.instance().getCroppedImg(validFile, crop, fileName);
    expect(spy).toHaveBeenCalled();
  });

  it('calls onCropChange function', () => {
    wrapper.instance().handleOnCropChange(crop);
    expect(wrapper.state().crop).toEqual(crop);
  });

  it('calls handleClose function', () => {
    const spy = jest.spyOn(wrapper.instance(), 'handleImageDrop');
    wrapper.instance().handleClose();
    wrapper.setState({ imageFile: validFile });
    expect(wrapper.state().src).toEqual('');
    expect(wrapper.state().open).toBeFalsy();
    wrapper.instance().handleImageDrop(wrapper.state().imageFile);
    expect(spy).toHaveBeenCalled();
  });

  it('calls handleSave function', () => {
    wrapper.instance().getCroppedImg = jest.fn(() => Promise.resolve(data));

    wrapper.instance().handleSave();
    wrapper.setState({
      src: 'image-src',
      fileName: 'eucerin',
      crop
    });
    expect(wrapper.instance().getCroppedImg).toHaveBeenCalled();
  });
  it('calls handleProposeSupplier function', () => {
    const newProps = {
      ...props,
      addSupplier: jest.fn(
        () => new Promise((resolve, reject) => reject({
          graphQLErrors: [
            {
              message: 'error'
            }
          ]
        }))
      )
    };
    wrapper.instance().handleProposeSupplier();
    wrapper.setProps(newProps);
  });
});
