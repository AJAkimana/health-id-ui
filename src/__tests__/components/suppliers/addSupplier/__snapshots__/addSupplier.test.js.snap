// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Add Supplier component renders data upon successful view 1`] = `ReactWrapper {}`;
